Document/0x05e-Testing-Cryptography.md

```diff
@@ -233,3 +233,52 @@ public static void main (String args[]) {
+
+A sligthly less secure way of storing encryption keys, is in the SharedPreferences of Android. When [SharedPreferences](https://developer.android.com/reference/android/content/SharedPreferences.html"Android SharedPreference API") are initialized in [MODE_PRIVATE](https://developer.android.com/reference/android/content/Context.html#MODE_PRIVATE, the file is only readable by the application that created it. However, on rooted devices any other application with root access can simply read the SharedPreference file of other apps, it does not matter whether MODE_PRIVATE has been used or not. This is not the case for the KeyStore. Since KeyStore access is managed on kernel level, which needs considerably more work and skill to bypass without the KeyStore clearing or destroying the keys. 
+
+The last two options are to use hardcoded encryption keys in the source code and storing generated keys in public places like ```/sdcard/```. Obviously, hardcoded encryption keys are not the way to go. This means every instance of the application uses the same encryption key. An attacker needs only to do the work once, to extract the key from the source code . Consequrently, he can decrypt any other data that he can obtain and that was encrypted by the application. Lastly, storing encryption keys publicly also is highly discouraged as other applications can have permission to read the public partition and steal the keys.
+
+#### Static Analysis
+
+Locate uses of the cryptographic primitives in reverse engineered or disassembled code. Some of the most frequently used classes and interfaces:
+
+- `Cipher`
+- `Mac`
+- `MessageDigest`
+- `Signature`
+- `KeyStore`
+- `Key`, `PrivateKey`, `PublicKey`, `SecretKey`
+- And a few others in the `java.security.*` and `javax.crypto.*` packages.
+
+Trace and assess how key material is generated and stored before or after it is used.
+
+#### Dynamic Analysis
+
+Hook cryptographic methods and analyze the keys that are being used. Monitor file system access while cryptographic operations are being performed to assess where key material is written to or read from.
+
+#### References
+
+##### OWASP MASVS
+- V3.1: "The app does not rely on symmetric cryptography with hardcoded keys as a sole method of encryption."
+- V3.5: "The app doesn't reuse the same cryptographic key for multiple purposes."
+##### OWASP Mobile Top 10 2016
+- M6 - Broken Cryptography
+
+##### CWE
+- CWE-321: Use of Hard-coded Cryptographic Key
+
+##### Info
+- [#nelenkov] - N. Elenkov, Android Security Internals, No Starch Press, 2014, Chapter 5.
```

Comment:
Please move this line directory under "References" and remove the "Info" section.